### PR TITLE
fix Issue 18053 - Use stdint.h mangling for int64_t/uint64_t when man…

### PR DESCRIPTION
**Pull request by WalterBright**
…gling D long/ulong

I'm tired of expending many hours dealing with this issue again and again. Fortunately, it only changes the mangling on OSX 64.

### CODE REVIEW

**Comment by WalterBright** [comment]
Note how much nicer the test cases get :-)

**Comment by WalterBright** [comment]
@ibuclaw @klickverbot please check effect on gdc/ldc

File: src/ddmd/cppmangle.d [hunk 1]
```diff
@@ -815,10 +815,12 @@ public:
             case Tuns32:                c = 'j';        break;
             case Tfloat32:              c = 'f';        break;
             case Tint64:
-                c = (Target.c_longsize == 8 ? 'l' : 'x');
+                // Mangle like target's stdint.h int64_t
+                c = (Target.c_longsize == 4 || global.params.isOSX ? 'x' : 'l');
```
Comment (ibuclaw) [inline src/ddmd/cppmangle.d]: This should be turned into a target hook. The frontend should not be asking questions such as "Am I targeting osx?". I think we already have a mangle type hook already present, the logic can be put there.

**Review by ibuclaw — DISMISSED** [review]

File: src/ddmd/cppmangle.d (hunk 1 above)
Comment (ibuclaw) [reply src/ddmd/cppmangle.d]: https://github.com/dlang/dmd/blob/5494dba193d16d6e126af8a12d37567b4d245086/src/ddmd/target.d#L459

Comment would need to be updated to reflect that if the target requires mangling in a specific way, then handle it here.

**Comment by jacob-carlborg** [comment]
I'm not sure which C++ compiler you're looking at but these are the results I got with Xcode 9.0 containing Clang `900.0.37`:

#### 64bit:

| Type               | C++ | D | D After PR |
|--------------------|-----|---|------------|
| `long`               | l   | l | x          |
| `long long`          | x   | l | x          |
| `int64_t`            | x   | l | x          |
| `unsigned long`      | m   | m | y          |
| `unsigned long long` | y   | m | y          |
| `uint64_t`           | y   | m | y          |

#### 32bit:

| Type               | C++ | D | D After PR |
|--------------------|-----|---|------------|
| `long`               | l   | i | i          |
| `long long`          | x   | x | x          |
| `int64_t`            | x   | x | x          |
| `unsigned long`      | m   | j | j          |
| `unsigned long long` | y   | y | y          |
| `uint64_t`           | y   | y | y          |

As can be seen in these tables, D uses the wrong mangling for the `long` type.

I've used the following C++ code:

```cpp
#include <stdint.h>

void a(long){}
void b(long long){}
void c(int64_t){}

void d(unsigned long){}
void e(unsigned long long){}
void f(uint64_t){}
```

And the following D code:

```d
extern (C++):

import core.stdc.config;
import core.stdc.stdint;

void a(c_long){}
void b(long){}
void c(int64_t){}

void d(c_ulong){}
void e(ulong){}
void f(uint64_t){}
```

**Comment by jacob-carlborg** [comment]
Hmm, seems like `__c_long/__c_ulong` needs to be used on macOS 64bit and not only 32bit.

**Comment by jacob-carlborg** [comment]
BTW, why this complicated logic with a struct with a specific name that matches the mangling of C++? Couldn't we make the following to work?

```d
pragma(mangle, "m") alias c_ulong = ulong;
```

**Comment by jacob-carlborg** [comment]
> I'd welcome a better solution for that. But not in this PR.

I have something that is a work in progress.

**Comment by JinShil** [comment]
@WalterBright Can you please rebase?

**Review by JinShil — APPROVED** [review]

**Comment by wilzbach** [comment]
This broke the OSX build on Travis, which shows the problems @jacob-carlborg pointed out. Should we revert, s.t. the error can be properly fixed/worked out?

**Comment by jacob-carlborg** [comment]
@wilzbach I have an alternative solution to the existing `__c_long` available here https://github.com/dlang/dmd/pull/7458. It adds support for specifying `pragma(mangle)` on aliases. Hopefully with this PR together with mine we could get the correct mangling without breaking any existing code.

**Pull request merged**